diff --git a/modules/avcodec/encode.c b/modules/avcodec/encode.c
index b45cd09..d9b4dd8 100644
--- a/modules/avcodec/encode.c
+++ b/modules/avcodec/encode.c
@@ -385,6 +385,7 @@ int avcodec_encode_update(struct videnc_state **vesp,
 	return err;
 }
 
+int packet_dup_handler(uint64_t ts, uint8_t *buf, size_t size, bool keyframe);
 
 int avcodec_encode(struct videnc_state *st, bool update,
 		   const struct vidframe *frame, uint64_t timestamp)
@@ -392,6 +393,7 @@ int avcodec_encode(struct videnc_state *st, bool update,
 	AVFrame *pict = NULL;
 	AVFrame *hw_frame = NULL;
 	AVPacket *pkt = NULL;
+	static bool update2 = false;
 	int i, err = 0, ret;
 	uint64_t ts;
 
@@ -443,10 +445,11 @@ int avcodec_encode(struct videnc_state *st, bool update,
 		pict->linesize[i] = frame->linesize[i];
 	}
 
-	if (update) {
+	if (update || update2) {
 		debug("avcodec: encoder picture update\n");
 		pict->key_frame = 1;
 		pict->pict_type = AV_PICTURE_TYPE_I;
+		update2 = false;
 	}
 
 	pict->color_range = AVCOL_RANGE_MPEG;
@@ -495,6 +498,10 @@ int avcodec_encode(struct videnc_state *st, bool update,
 
 	ts = video_calc_rtp_timestamp_fix(pkt->pts);
 
+	ret = packet_dup_handler(timestamp, pkt->data, pkt->size, !!(pkt->flags & AV_PKT_FLAG_KEY));
+	if (ret)
+		update2 = true;
+
 	switch (st->codec_id) {
 
 	case AV_CODEC_ID_H264:
