diff --git a/src/video.c b/src/video.c
index eec6a35c..353019ca 100644
--- a/src/video.c
+++ b/src/video.c
@@ -24,7 +24,7 @@
 /** Video transmit parameters */
 enum {
 	MEDIA_POLL_RATE = 250,		       /**< in [Hz]                  */
-	BURST_MAX	= 8192,		       /**< in bytes                 */
+	BURST_MAX	= 32768,	       /**< in bytes                 */
 	RTP_PRESZ	= 4 + RTP_HEADER_SIZE, /**< TURN and RTP header      */
 	RTP_TRAILSZ	= 12 + 4,	       /**< SRTP/SRTCP trailer       */
 	PICUP_INTERVAL	= 500,		       /**< FIR/PLI interval         */
@@ -292,7 +292,7 @@ static void vidqueue_poll(struct vtx *vtx, uint64_t jfs, uint64_t prev_jfs)
 	bandwidth_kbps = vtx->video->cfg.bitrate / 1000;
 	burst = (size_t)((1 + jfs - prev_jfs) * bandwidth_kbps / 4);
 
-	burst = min(burst, BURST_MAX);
+	burst = max(burst, BURST_MAX);
 	sent  = 0;
 
 	while (le) {
