diff --git a/src/video.c b/src/video.c
index eec6a35c..f3371311 100644
--- a/src/video.c
+++ b/src/video.c
@@ -449,7 +449,6 @@ static void encode_rtp_send(struct vtx *vtx, struct vidframe *frame,
 {
 	struct le *le;
 	int err = 0;
-	bool sendq_empty;
 
 	if (!vtx->enc)
 		return;
@@ -471,15 +470,6 @@ static void encode_rtp_send(struct vtx *vtx, struct vidframe *frame,
 		goto out;
 	}
 
-	mtx_lock(&vtx->lock_tx);
-	sendq_empty = (vtx->sendq.head == NULL);
-	mtx_unlock(&vtx->lock_tx);
-
-	if (!sendq_empty) {
-		++vtx->skipc;
-		return;
-	}
-
 	mtx_lock(&vtx->lock_enc);
 
 	/* Convert image */
