diff --git a/include/baresip.h b/include/baresip.h
index ca60de75..e32da758 100644
--- a/include/baresip.h
+++ b/include/baresip.h
@@ -1429,6 +1429,7 @@ const char *stream_peer(const struct stream *strm);
 int  stream_bundle_init(struct stream *strm, bool offerer);
 int  stream_debug(struct re_printf *pf, const struct stream *s);
 void stream_enable_rtp_timeout(struct stream *strm, uint32_t timeout_ms);
+void stream_flush(struct stream *s);
 
 
 /*
diff --git a/src/core.h b/src/core.h
index 2b45f069..759b3de0 100644
--- a/src/core.h
+++ b/src/core.h
@@ -298,7 +298,6 @@ int  stream_resend(struct stream *s, uint16_t seq, bool ext, bool marker,
 		  int pt, uint32_t ts, struct mbuf *mb);
 
 /* Receive */
-void stream_flush(struct stream *s);
 int  stream_decode(struct stream *s);
 int  stream_ssrc_rx(const struct stream *strm, uint32_t *ssrc);
 
diff --git a/src/stream.c b/src/stream.c
index 57c832e2..727e09c8 100644
--- a/src/stream.c
+++ b/src/stream.c
@@ -354,7 +354,7 @@ static void rtp_handler(const struct sa *src, const struct rtp_header *hdr,
 
 	MAGIC_CHECK(s);
 
-	if (!s->rx.enabled && s->type == MEDIA_AUDIO)
+	if (!s->rx.enabled)
 		return;
 
 	if (rtp_pt_is_rtcp(hdr->pt)) {
@@ -1167,8 +1167,7 @@ void stream_flush(struct stream *s)
 	if (s->rx.jbuf)
 		jbuf_flush(s->rx.jbuf);
 
-	if (s->type == MEDIA_AUDIO)
-		rtp_clear(s->rtp);
+	rtp_clear(s->rtp);
 }
 
 
