diff --git a/include/baresip.h b/include/baresip.h
index 520a968d..33904e7e 100644
--- a/include/baresip.h
+++ b/include/baresip.h
@@ -1641,6 +1641,8 @@ int  mediatrack_start_audio(struct media_track *media,
 			    struct list *ausrcl, struct list *aufiltl);
 int  mediatrack_start_video(struct media_track *media);
 struct stream *media_get_stream(const struct media_track *media);
+struct audio *media_get_audio(const struct media_track *media);
+struct video *media_get_video(const struct media_track *media);
 enum media_kind mediatrack_kind(const struct media_track *media);
 const char *media_kind_name(enum media_kind kind);
 
diff --git a/src/mediatrack.c b/src/mediatrack.c
index 1373eddc..2cc90ee0 100644
--- a/src/mediatrack.c
+++ b/src/mediatrack.c
@@ -167,6 +167,24 @@ struct stream *media_get_stream(const struct media_track *media)
 }
 
 
+struct audio *media_get_audio(const struct media_track *media)
+{
+	if (!media || media->kind != MEDIA_KIND_AUDIO)
+		return NULL;
+
+	return media->u.au;
+}
+
+
+struct video *media_get_video(const struct media_track *media)
+{
+	if (!media || media->kind != MEDIA_KIND_VIDEO)
+		return NULL;
+
+	return media->u.vid;
+}
+
+
 const char *media_kind_name(enum media_kind kind)
 {
 	switch (kind) {
